### ISSUE TITLE
About time consuming for "after training"

### ISSUE THREAD

**Issue by planemanner** [ganguli-lab/Synaptic-Flow#8]
Hi, thanks for your great effort on this research.

First of all, please note that I'm a newbie in pruning.

I have a question for time consumption when  train a pruned model with data.

Following is my shellscript for experiment.

declare -a PrunerArray=("synflow")

for seed in 0 1 2; do
	for val in "${PrunerArray[@]}"; do
		for d in 0.00 0.25 0.50 0.75 1.00 1.25 1.50 1.75 2.00 2.25 2.50 2.75 3.00; do
			if [ "$val" = "synflow" ]
			then
				CUDA_VISIBLE_DEVICES=3 python3 main.py --experiment=singleshot --optimizer=momentum --lr=0.1 --lr-drops 60 120 --lr-drop-rate=0.1 --train-batch-size=128 --weight-decay=1e-4 --pre-epochs=0 --post-epochs=160 --gpu=0 --result-dir=./Results --expid=cifar10_lottery_vgg16_${val}comp_ratio${d}run${seed} --seed=$seed --dataset=cifar10 --model-class=lottery --model=vgg16 --pruner=$val --prune-epochs=100 --compression=$d
			else
				CUDA_VISIBLE_DEVICES=3 python3 main.py --experiment=singleshot --optimizer=momentum --lr=0.1 --lr-drops 60 120 --lr-drop-rate=0.1 --train-batch-size=128 --weight-decay=1e-4 --pre-epochs=0 --post-epochs=160 --gpu=0 --result-dir=./Results --expid=cifar10_lottery_vgg11_bn${val}comp_ratio${d}_run${seed} --seed=$seed --dataset=cifar10 --model-class=lottery --model=vgg11-bn --prune-train-mode=True --pruner=$val --prune-epochs=1 --compression=$d
			fi
		done
	done
done

After getting pruned models, I checked post-training times for each compression ratios.
But, I got weired results. 

In case of compression ratio : 0.00
Expected time : 24 min 41 sec

In case of compression ratio : 0.25
Expected time : 23 min 41 sec

In case of compression ratio : 0.5
Expected time : 24 min 21 sec
...

Although I'm using a bare metal with my peers, considering compression ratio, I think it should conspicuously take much lower time for high compression ratio.  

Could you answer what is wrong within ?

Should I do remove pruning re-parametrization by reffering [Pruning Tutorial](https://pytorch.org/tutorials/intermediate/pruning_tutorial.html)?

------------------------------------------------------------------------------

I've got an answer from [here](https://stackoverflow.com/questions/62326683/prunning-model-doesnt-improve-inference-speed-or-reduce-model-size)

For a person who would have a similar question in future, I briefly leave an answer. 

Since the "synflow" is a kind of unconstructed pruning method, we can only make the weights having 0 values be removed in case of structed one by using Tensor.to_sparse().

For example,

```python
a = torch.tensor([[0, 0, 1, 0], [1, 2, 0, 0], [0, 0, 0, 0]], dtype=torch.float64)
print(a)
sp = a.to_sparse()
print(sp)
```

<img width="387" alt="image" src="https://user-images.githubusercontent.com/59157285/123044360-80400800-d434-11eb-8657-e53d54ce691c.png">
